File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/sender/impl/RocketSender.java
package org.xyattic.eventual.consistency.support.core.sender.impl;

import lombok.extern.slf4j.Slf4j;
import org.apache.rocketmq.client.producer.SendResult;
import org.apache.rocketmq.client.producer.SendStatus;
import org.apache.rocketmq.spring.core.RocketMQTemplate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.messaging.support.GenericMessage;
import org.springframework.transaction.annotation.Transactional;
import org.xyattic.eventual.consistency.support.core.persistence.ProviderPersistence;
import org.xyattic.eventual.consistency.support.core.provider.PendingMessage;
import org.xyattic.eventual.consistency.support.core.provider.enums.PendingMessageStatus;
import org.xyattic.eventual.consistency.support.core.sender.Sender;

/**
 * @author wangxing
 * @create 2020/6/29
 */
@Slf4j
public class RocketSender implements Sender {

    @Autowired
    private RocketMQTemplate rocketMQTemplate;
    @Autowired
    private ProviderPersistence providerPersistence;

    @Override
    @Transactional(rollbackFor = Exception.class)
    public void send(PendingMessage pendingMessage) {
        SendResult sendResult = rocketMQTemplate.syncSend(pendingMessage.getDestination(),
                new GenericMessage<>(pendingMessage.getBody(), pendingMessage.getHeaders()));
        if (SendStatus.SEND_OK.equals(sendResult.getSendStatus())) {
            log.info("发送成功:{}", sendResult);
            providerPersistence.changePendingMessageStatus(pendingMessage.getMessageId(),
                    PendingMessageStatus.HAS_BEEN_SENT);
        } else {
            log.warn("发送失败:" + sendResult);
        }
    }

}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/sender/impl/RocketSender.kt
package org.xyattic.eventual.consistency.support.core.sender.impl

import org.apache.rocketmq.client.producer.SendStatus
import org.apache.rocketmq.spring.core.RocketMQTemplate
import org.apache.rocketmq.spring.support.RocketMQHeaders
import org.springframework.messaging.support.GenericMessage
import org.xyattic.eventual.consistency.support.core.provider.PendingMessage
import org.xyattic.eventual.consistency.support.core.provider.PendingMessageHeaders
import org.xyattic.eventual.consistency.support.core.utils.getLogger

/**
 * @author wangxing
 * @create 2020/6/29
 */
open class RocketSender(private val rocketMQTemplate: RocketMQTemplate) : AbstractSender() {

    companion object {
        private val log = getLogger()
    }


    override fun send(pendingMessage: PendingMessage) {
        getTransactionTemplate(pendingMessage)
            .execute {
                var topicWithTags: String = pendingMessage.destination
                if (pendingMessage.headers.containsKey(RocketMQHeaders.TAGS)) {
                    topicWithTags += ":${pendingMessage.headers[RocketMQHeaders.TAGS]}"
                }
                val sendResult =
                    if (pendingMessage.headers.containsKey(PendingMessageHeaders.partitionHeader)) {
                        val partitionKey =
                            pendingMessage.headers[PendingMessageHeaders.partitionHeader].toString()
                        log.info("send orderly message with '$partitionKey' for $pendingMessage")
                        rocketMQTemplate.syncSendOrderly(
                            topicWithTags,
                            GenericMessage(pendingMessage.body, pendingMessage.headers),
                            partitionKey
                        )
                    } else {
                        rocketMQTemplate.syncSend(
                            topicWithTags,
                            GenericMessage(pendingMessage.body, pendingMessage.headers)
                        )
                    }
                if (SendStatus.SEND_OK == sendResult.sendStatus) {
                    log.info("sent successfully: ${sendResult}")
                    getProviderPersistence(pendingMessage.persistenceName).sendSuccess(
                        pendingMessage.messageId,
                        sendResult.msgId
                    )
                } else {
                    log.warn("failed to send: $sendResult")
                }
            }

    }

}